tools/traffic/table/blob_table.go

```diff
@@ -0,0 +1,125 @@
+package table
+
+import (
+	"fmt"
+	"math/rand"
+	"sync"
+)
+
+// BlobTable tracks blobs written by the traffic generator. This is a thread safe data structure.
+type BlobTable struct {
+
+	// blobs contains all blobs currently tracked by the requiredReads.
+	blobs []*BlobMetadata
+
+	// size describes the total number of blobs currently tracked by the requiredReads.
+	// size may be smaller than the capacity of the blobs slice.
+	size uint
+
+	// lock is used to synchronize access to the requiredReads.
+	lock sync.Mutex
+}
+
+// NewBlobTable creates a new BlobTable instance.
+func NewBlobTable() BlobTable {
+	return BlobTable{
+		blobs: make([]*BlobMetadata, 1024),
+		size:  0,
+	}
+}
+
+// Size returns the total number of blobs currently tracked by the requiredReads.
+func (table *BlobTable) Size() uint {
+	table.lock.Lock()
+	defer table.lock.Unlock()
+
+	return table.size
+}
+
+// Get returns the blob at the specified index. Returns nil if the index is out of bounds.
+func (table *BlobTable) Get(index uint) *BlobMetadata {
+	table.lock.Lock()
+	defer table.lock.Unlock()
+
+	if index >= table.size {
+		return nil
+	}
+
+	return table.blobs[index]
+}
+
+// Add a blob to the requiredReads.
+func (table *BlobTable) Add(blob *BlobMetadata) {
+	table.lock.Lock()
+	defer table.lock.Unlock()
+
+	blob.index = table.size
+	table.blobs[table.size] = blob
+	table.size++
+}
+
+// AddOrReplace adds a blob to the requiredReads if there is capacity or replaces an existing blob at random
+// if the requiredReads is full. This method is a no-op if maximumCapacity is 0.
+func (table *BlobTable) AddOrReplace(blob *BlobMetadata, maximumCapacity uint) {
+	if maximumCapacity == 0 {
+		return
+	}
+
+	table.lock.Lock()
+	defer table.lock.Unlock()
+
+	if table.size >= maximumCapacity {
+		// replace random existing blob
```

Comment:
what if the existing blob hasn't been retrieved the required number of times?

Reply:
Then it is removed.

This is not a problem for our use case though, since we keep two blob tables: one for required blobs, and another for optional blobs. The code never calls `AddOrReplace()` on the table that contains the required blobs.